src/cluster_distances.jl

```diff
@@ -0,0 +1,233 @@
+                          batch_size::Union{Integer, Nothing} = nothing)
+    (batch_size === nothing) || (size(points, 2) > batch_size) ||
+        error("batch-updates of distance matrix-based ClusterDistances not supported")
+    SimpleClusterDistances(metric, assignments, points)
+end
+
+Base.isready(dists::SimpleClusterDistances) = length(dists.assignments) == size(dists.point_dists, 1)
+
+function update!(dists::SimpleClusterDistances{M},
+                 assignments::AbstractVector{<:Integer},
+                 points::AbstractMatrix{<:Real}) where M
+    @assert length(assignments) == size(points, 2)
+    check_assignments(assignments, nclusters(dists))
+    append!(dists.assignments, assignments)
+    length(dists.assignments) == size(dists.point_dists, 1) ||
+        error("$(typeof(dists)) does not support batch updates: $(length(assignments)) points given, $(size(dists.point_dists, 1)) expected))")
+    @inbounds for cluster in assignments
+        dists.cluster_sizes[cluster] += 1
+    end
+
+    n = size(dists.point_dists, 1)
+    if M === Nothing
+        @assert M === Nothing
+        @assert length(dists.assignments) == n
+        @assert size(point_dists) == (n, n)
+        copy!(dists.point_dists, point_dists)
+    else
+        # metric-based SimpleClusterDistances does not support batched updates
+        @assert size(points, 2) == n
+        pairwise!(dists.metric, dists.point_dists, points, dims=2)
+    end
+
+    return dists
+end
+
+# this function returns matrix r nclusters x n, such that
+# r[i, j] is the sum of distances of all points from cluster i to point j
+function sumdistances(dists::SimpleClusterDistances,
+                      points::Union{AbstractMatrix, Nothing},
+                      indices::AbstractVector{<:Integer})
+    @assert isready(dists)
+    T = eltype(dists.point_dists)
+    n = length(dists.assignments)
+    S = typeof((one(T)+one(T))/2)
+    r = zeros(S, nclusters(dists), n)
+    @inbounds for j in indices
+        for i = 1:j-1
+            r[dists.assignments[i], j] += dists.point_dists[i,j]
+        end
+        for i = j+1:n
+            r[dists.assignments[i], j] += dists.point_dists[i,j]
+        end
+    end
+    return r
+end
+
+# uses the method from "Distributed Silhouette Algorithm: Evaluating Clustering on Big Data"
+# https://arxiv.org/abs/2303.14102
+# for SqEuclidean point distances
+struct SqEuclideanClusterDistances{T} <: ClusterDistances{T}
+    npoints::Int
+    cluster_sizes::Vector{Int} # [nclusters]
+    Y::Matrix{T} # [dims, nclusters], the first moments of each cluster (sum of point coords)
+    Ψ::Vector{T} # [nclusters], the second moments of each cluster (sum of point coord squares)
+
+    SqEuclideanClusterDistances(::Type{T}, npoints::Integer, dims::Integer,
+                                nclusters::Integer) where T<:Real =
+        new{T}(npoints, zeros(Int, nclusters), zeros(T, dims, nclusters), zeros(T, nclusters))

```

Comment:
`npoints` is set here and is never updated during `update!`
This is fine for small datasets, but imagine cases of
1. a streaming data source where the number of points is not known in advance. In particular the points used for calculating silhouettes is not the same as the number of points that were used for training.
2. The dataset is so big that we can actually calculate in a streaming fashion until silhouette scores converge, and we stop the calculation for actual `npoints` less than this predefined value.

In short, `npoints` should be updated along with `cluster_sizes` on line 218.